cpp/include/rapidsmpf/shuffler/chunk.hpp

```diff
@@ -24,6 +26,291 @@ namespace rapidsmpf::shuffler::detail {
  */
 using ChunkID = std::uint64_t;
 
+/**
+ * @brief Chunk with multiple messages.
+ *
+ * This class will have two buffers:
+ * - metadata_: The metadata buffer that contains information about the messages in the
+ * chunks and the concatenated metadata of the messages.
+ * - data_: The data buffer that contains the concatenateddata of the messages in the
+ * chunk.
+ *
+ * The metadata_ buffer will have the following format:
+ * - chunk_id: uint64_t, ID of the chunk
+ * - n_elements: size_t, Number of messages in the chunk
+ * - [partition_ids]: std::vector<PartID>, Partition IDs of the messages, size =
+ * n_elements
+ * - [expected_num_chunks]: std::vector<size_t>, Expected number of chunks of the
+ * messages, size = n_elements
+ * - [psum_meta]: std::vector<uint32_t>, Prefix sums (excluding 0) of the metadata
+ * sizes of the messages, size = n_elements
+ * - [psum_data]: std::vector<uint64_t>, Prefix sums (excluding 0) of the data sizes of
+ * the messages, size = n_elements
+ * - [concat_metadata]: std::vector<uint8_t>, Concatenated metadata of the messages,
+ * size = psum_meta[n_elements - 1]
+ *
+ * For a chunk with N messages with M bytes of concat metadata the size of metadata_
+ * buffer is sizeof(ChunkID) + sizeof(size_t) + N * (sizeof(PartID) + sizeof(size_t) +
+ * sizeof(uint32_t) + sizeof(uint64_t)) + M = 16 + N * 24 + M bytes.
+ *
+ * For a chunk with a single control message, the size of the metadata_ buffer is
+ * sizeof(ChunkID) + sizeof(PartID)+ 2*sizeof(size_t) + sizeof(uint32_t) +
+ * sizeof(uint64_t) = 40 bytes.
+ *
+ * For a chunk with a single message with M bytes of metadata, the size of the metadata_
+ * buffer is sizeof(ChunkID) + sizeof(PartID) + sizeof(size_t) + sizeof(uint32_t) +
+ * sizeof(ChunkID) + sizeof(PartID) + sizeof(size_t) + sizeof(uint32_t) + sizeof(uint64_t)
+ * + M = 40 + M bytes.
+ */
+
+// TODO: This class will be renamed to `Chunk` once the `Chunk` class is removed.
+class ChunkBatch {
+  public:
+    /**
+     * @brief The size of the metadata message header.
+     *
+     * @param n_messages The number of messages in the chunk.
+     * @return The size of the metadata message header.
+     */
+    static constexpr size_t metadata_message_header_size(size_t n_messages) {
+        return sizeof(ChunkID) + sizeof(size_t)
+               + n_messages
+                     * (sizeof(PartID) + sizeof(size_t) + sizeof(uint32_t)
+                        + sizeof(uint64_t));
+    }
+
+    /**
+     * @brief The ID of the chunk.
+     *
+     * @return The ID of the chunk.
+     */
+    inline ChunkID chunk_id() const {
+        return *reinterpret_cast<ChunkID*>(metadata_->data());
+    }
+
+    /**
+     * @brief The number of messages in the chunk.
+     *
+     * @return The number of messages in the chunk.
+     */
+    inline size_t n_messages() const {
+        return *reinterpret_cast<size_t*>(metadata_->data() + sizeof(ChunkID));
```

Comment:
All these reinterpret_cast type-punning approaches break strict-aliasing rules unfortunately.

What we should do is:

1. (C++ 20) use `std::bit_cast` (but it's messy because we're carrying around `std::byte` and `bit_cast` takes values, not pointer + size).
2. Use `memcpy` (the compiler will optimise this)

So, this would be (for example):

```
inline size_t n_messages() const {
    size_t result;
    memcpy(&result, metadata_->data() + sizeof(ChunkID), sizeof(result));
    return result;
}
```